### ISSUE TITLE
Optional.new([1, 2, 3]) - calling .first gives me the whole array, calling .last gives me Optional 3

### ISSUE THREAD

**Issue by KurtRMueller** [tomstuart/monads#4]
```ruby
Optional.new([1, 2, 3]).first # => [1, 2, 3]
Optional.new([1, 2, 3]).last # => #<struct Monads::Optional value=3>

Optional.new([1, 2, 3]).last(2) # => #<struct Monads::Optional value=[2, 3]>
Optional.new([1, 2, 3]).first(2) # => [[1, 2, 3]]
```
I'm not knowledgable enough about the behavior of the Maybe monad to tell if this is the correct behavior or if this is a bug.

**Comment by MMF2** [comment]
I've stumbled upon a similar problem. It is a symptom of a slightly deeper problem.
Explanation of what happens first:
Optional is a class generated from Struct. As such it mixes in the Enumerable module within which the "first" but not the "last" method is defined. A Struct being considered an Enumerable means it is considered a collection of its attributes. Optional has one attribute (value).

The general problem is that only the call to methods not defined within the Option class are forwarded to the Option's value. Exactly which methods are defined on Option is somewhat surprising and a potential source of bugs. In this case "first" is defined on Option, while "last" is not.

I think the user expects essentially all method calls on an Optional to be forwarded to its value. Exceptions, if necessary, should be documented explicitely.

Some other methods I expected on a first naive account to be forwarded to an Option's value, but which are not, are e.g.
==, [], nil?

Suggestions: 
1. Reimplement Optional (and possibly the other monads, I have not considered them yet) as a normal class rather than one generated by Struct (it has only one attribute, anyway). This way we get rid of Optional's own disruptive implementation of, e.g., "[]" which is not needed anyway and all the methods that come with Enumerable (as e.g. "first") which are not needed either since Optional is a degenerated collection of a single attribute only.
2. Have a look on all inherited methods when Optional is made a subclass of Object. Either reimplement them such that they are explicitely forwarded to the Optional's value (as it would make sense, e.g., I would say for nil?) or document explicitely that they are not forwarded.

**Comment by tomstuart** [comment]
Sorry, I had notifications turned off, so only just saw this conversation when someone pointed me to it.

You’re (both) right that this is unwanted functionality introduced by `Struct`. I suppose the question is whether it’s worth fixing it up by reimplementing each monad as inheriting from (probably) `BasicObject` so that we don’t get any undesirable methods.

My main concern with changing this (or anything else) is that this code is meant to be educational rather than production-ready, so I’m hesitant to do anything too unusual for the sake of actual usability or performance. Then again, maybe the code would actually be improved by cleaning out `Struct` — it’s not really pulling its weight here anyway.

WDYT?

**Comment by MMF2** [comment]
To my mind, the label "educational" justifies low performance but not unexpected behavior. That some methods like, say, "first" are not forwarded to an Optional's value has nothing to do with performance but could be considered something similar to a bug since this behavior does not match the expectations for a Maybe monad. I still recommend to reimplement at least the Optional class as a normal subclass of Object. I do not see any connection between the basic algorithmic idea (explained very convincing as a series of refactorings at http://codon.com/refactoring-ruby-with-monads, thank you!) and the implementation of Optional as a Struct. Therefore, I would see no loss in the educational quality with a Struct-free reimplementation (it's a few minutes, I am just doing it). If you think on it, you probably admit that implementing Optional as a Struct had no good reason in the first place. On the other hand, making Optional a direct subclass of BasicObject would possibly entail additional complications obscuring the basic idea and therefore might destroy the educational flavor (just a guess, have not seriously considered the implications). To sum up, I think making Optional a subclass of Object would be worth it.
By the way: If you make a new version anyway, I suggest in addition to replace the public_send in Monad#method_missing by a simple send since I can see no justification for the limitation to public methods and - at least to consider - to forward the nil? method and to reimplement the == method. Thank you.

**Comment by marcinwyszynski** [comment]
https://github.com/tomstuart/monads/pull/7

**Comment by tomstuart** [comment]
Fixed by #10.
